### PR TITLE
[4.0] Replace "label" classes with the new "badge" ones

### PR DESCRIPTION
**Pull request by htmgarcia**
Pull Request for Issue # .

### Summary of Changes
Replace "label" classes (not available in Bootstrap 4) with the new "badge" ones in frontend and backend

### Testing Instructions
You can test in several places. 
As example. Preview single articles that include tags in frontend 

### Expected result
The tags should have styling (color, background, etc) (with "badge" classes)

### Actual result
The tags looks like regular links (with "label" classes)

### Documentation Changes Required
Replacing the "label" classes with "badge" ones. That summarize this PR.

### CODE REVIEW

**Comment by brianteeman** [comment]
Can you please check in this repo where the admin template etc is being developed as I believe this has already been done https://github.com/joomla/40-backend-template/

**Comment by htmgarcia** [comment]
@brianteeman I didn't included any change related to the backend template.

File: components/com_content/tmpl/category/default_articles.php [hunk 1]
```diff
@@ -186,24 +186,24 @@
...
-								<?php $class = 'label label-association label-' . $association['language']->sef; ?>
+								<?php $class = 'badge badge-association badge-' . $association['language']->sef; ?>
 								&nbsp;<a class="' . <?php echo $class; ?> . '" href="<?php echo JRoute::_($association['item']); ?>"><?php echo strtoupper($association['language']->sef); ?></a>&nbsp;
 							<?php endif; ?>
 						<?php endforeach; ?>
 					<?php endif; ?>
 				<?php endif; ?>
 				<?php if ($article->state == 0) : ?>
-					<span class="list-published label label-warning">
+					<span class="list-published label badge-warning">
```
Comment (Quy) [inline components/com_content/tmpl/category/default_articles.php]: Replace

File: components/com_content/tmpl/category/default_articles.php [hunk 2]
```diff
@@ -186,24 +186,24 @@
...
 						<?php echo JText::_('JUNPUBLISHED'); ?>
 					</span>
 				<?php endif; ?>
 				<?php if (strtotime($article->publish_up) > strtotime(JFactory::getDate())) : ?>
-					<span class="list-published label label-warning">
+					<span class="list-published label badge-warning">
```
Comment (Quy) [inline components/com_content/tmpl/category/default_articles.php]: Replace

File: components/com_content/tmpl/category/default_articles.php [hunk 3]
```diff
@@ -186,24 +186,24 @@
...
 						<?php echo JText::_('JNOTPUBLISHEDYET'); ?>
 					</span>
 				<?php endif; ?>
 				<?php if ((strtotime($article->publish_down) < strtotime(JFactory::getDate())) && $article->publish_down != JFactory::getDbo()->getNullDate()) : ?>
-					<span class="list-published label label-warning">
+					<span class="list-published label badge-warning">
```
Comment (Quy) [inline components/com_content/tmpl/category/default_articles.php]: Replace

File: libraries/cms/html/select.php [hunk 4]
```diff
@@ -362,8 +362,8 @@ public static function integerlist($start, $end, $inc, $name, $attribs = null, $
 	 *                             option.key: The property that will hold the selection value.
 	 *                             Defaults to "value".
 	 *                             option.label: The property in each option array to use as the
-	 *                             selection label attribute. If a "label" option is provided, defaults to
-	 *                             "label", if no label is given, defaults to null (none).
+	 *                             selection label attribute. If a "badge" option is provided, defaults to
```
Comment (Quy) [inline libraries/cms/html/select.php]: Are you sure these changes are applicable here?

File: components/com_content/tmpl/category/default_articles.php (hunk 3 above)
Comment (htmgarcia) [reply components/com_content/tmpl/category/default_articles.php]: Done

File: libraries/cms/html/select.php (hunk 4 above)
Comment (htmgarcia) [reply libraries/cms/html/select.php]: Reverted

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on b8c7f7f54caafedd41a78045cb120851e161e5ff<br /><br />Code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20328">issues.joomla.org/tracker/joomla-cms/20328</a>.</sub>

**Comment by htmgarcia** [comment]
@Quy I don't see the problem with the conflicting files. Any idea?

**Comment by htmgarcia** [comment]
Resolved conflicts.

**Comment by htmgarcia** [comment]
@rdeutz everything looks fine to me. May you review this PR, please?

**Comment by dgrammatiko** [comment]
@ciar4n and @wilsonge  do you really want to tightly couple the front end to Bootstrap? There is always the option to extend to some random unique class which will work nicely with but especially without Bootstrap (you can even keep the old classes and extend them to badge).

Personally I wouldn't tightly couple the **FRONTEND**  core output to Bootstrap, it brings more limitations than the promised easiness of doing things. Also it's a no go if someone needs to build a fairly performant site, whatever that means according to the current benchmarking tools out there (google is probably installed in every client pc, so lighthouse is just a click away). Food for thought...

**Comment by mbabker** [comment]
The badge/label utility class isn't going to "tightly couple the **FRONTEND**  core output to Bootstrap".

- https://semantic-ui.com/elements/label.html
- https://foundation.zurb.com/sites/docs/badge.html or https://foundation.zurb.com/sites/docs/label.html
- https://materializecss.com/badges.html
- http://getbootstrap.com/docs/4.1/components/badge/

Pick a standard, stick with it, or say "screw every existing standard" and write the Joomla CSS Framework already.

**Comment by brianteeman** [comment]
@htmgarcia it can't be merged until it passes all the tests

**Comment by htmgarcia** [comment]
@brianteeman This PR is just about replacing "label" as classname in the HTML. Nothing fancy. 
Sorry. I don't follow the 2 fails on the checks. Any idea what's the issue with these failed checks?

**Comment by ciar4n** [comment]
At this stage it is safe to say we are not going to be creating our own CSS framework. Prefixing one of the current standards (eg Bootstrap) is the only viable option **IF** we want to decouple the views from Bootstrap. To do that we either

1. Rewrite an existing framework manually, taking what we need and prefixing all the classes (#15225)
2. Automating the prefixing at a SCSS compiler level.

I've looked in to the second option (the ideal of the 2) however the automated prefixing modules I've tested to date do not play nice with SCSS `@extend`s which pretty much rules out using Bootstrap. Most other frameworks avoid `@extend` (https://csswizardry.com/2016/02/mixins-better-for-performance/).

**Comment by htmgarcia** [comment]
@brianteeman @wilsonge may we merge this one, pretty please?

**Comment by brianteeman** [comment]
@htmgarcia it can't be merged until it passes all the tests

it still has zero user tests

**Comment by htmgarcia** [comment]
@brianteeman https://github.com/joomla/joomla-cms/pull/20328#issuecomment-388096710

**Comment by brianteeman** [comment]
tests are reset when the code changes which you did when you resolved the conflicts

**Comment by Quy** [comment]
Found some more. Please see screenshot.

badge

**Comment by Quy** [comment]
There are 3 more in the screenshot that have to be replaced.

File: administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php [hunk 5]
```diff
@@ -29,7 +29,7 @@
 						<?php echo $option->label; ?>
 					</td>
 					<td>
-							<span class="label label-<?php echo $option->state ? 'success' : 'important'; ?>">
+							<span class="badge badge-<?php echo $option->state ? 'success' : 'important'; ?>">
```
Comment (Quy) [inline administrator/components/com_joomlaupdate/tmpl/joomlaupdate/default_preupdatecheck.php]: `label-important` is now `badge-danger`.

**Comment by Quy** [comment]
Sorry found more. These should be all.

\administrator\language\en-GB\en-GB.com_joomlaupdate.ini

```
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_BREAK="*Extensions marked with <span class='label label-important'>No</span> or <span class='label'>Missing Compatibility Tag</span> might break your website. Please consult with the developer before upgrading."
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_MISSING_TAG="*Extensions marked with <span class='label'>Missing Compatibility Tag</span> indicate the developer did not add a compatibility tag in the respective extension's XML."
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_UPDATE_REQUIRED="*Extensions marked with <span class='label label-warning'>Yes (X.X.X)</span> might require an update."
```

**Comment by Quy** [comment]
Sorry these should be the last ones.

```
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_BREAK="*Extensions marked with <span class='label label-important'>No</span> or <span class='label'>Missing Compatibility Tag</span> might break your website. Please consult with the developer before upgrading."
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_MISSING_TAG="*Extensions marked with <span class='label'>Missing Compatibility Tag</span> indicate the developer did not add a compatibility tag in the respective extension's XML."
COM_JOOMLAUPDATE_VIEW_DEFAULT_DESCRIPTION_UPDATE_REQUIRED="*Extensions marked with <span class='label label-warning'>Yes (X.X.X)</span> might require an update."
```

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 59ed416bd31d86482ed0ecb878b196a383ef323f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20328">issues.joomla.org/tracker/joomla-cms/20328</a>.</sub>

**Comment by wilsonge** [comment]
Nice work! Thanks :)

**Pull request merged**